tools/cell_census_builder/validate.py

```diff
@@ -281,30 +305,29 @@ def _validate_X_layers_contents_by_dataset(args: Tuple[str, str, Dataset, List[E
                 .soma_joinid.to_numpy()
             )
 
+            raw_sum = 0.0
             raw_nnz = 0
-            if len(soma_joinids) > 0:
-                assert soma.SparseNDArray.exists(exp.ms["RNA"].X["raw"].uri)
-
-                def count_elements(arr: soma.SparseNDArray, join_ids: npt.NDArray[np.int64]) -> int:
-                    return sum(t.non_zero_length for t in arr.read((join_ids, slice(None))).coos())
-
-                raw_nnz = count_elements(exp.ms["RNA"].X["raw"], soma_joinids)
-
-            def count_nonzero(arr: Union[sparse.spmatrix, npt.NDArray[Any]]) -> int:
-                """Return _actual_ non-zero count, NOT the stored value count."""
-                if isinstance(arr, (sparse.spmatrix, sparse.coo_array, sparse.csr_array, sparse.csc_array)):
-                    return np.count_nonzero(arr.data)
-                return np.count_nonzero(arr)
+            fdpm_gene_count = 0
+            if len(cell_soma_joinids) > 0:
+                assert soma.SparseNDArray.exists(exp.ms[MEASUREMENT_RNA_NAME].X["raw"].uri)
+                raw_nnz = _count_elements(exp.ms[MEASUREMENT_RNA_NAME].X["raw"], cell_soma_joinids)
+                raw_sum = _sum_elements(exp.ms[MEASUREMENT_RNA_NAME].X["raw"], cell_soma_joinids)
+                fdpm_gene_count = _count_elements(
+                    exp.ms["RNA"][FEATURE_DATASET_PRESENCE_MATRIX_NAME], [dataset.soma_joinid]
+                )
 
             if ad.raw is None:
-                assert raw_nnz == count_nonzero(
-                    ad.X
-                ), f"{eb.name}:{dataset.dataset_id} 'raw' nnz mismatch {raw_nnz} vs {count_nonzero(ad.X)}"
+                ad_x = ad.X
             else:
-                assert raw_nnz == count_nonzero(
-                    ad.raw.X
-                ), f"{eb.name}:{dataset.dataset_id} 'raw' nnz mismatch {raw_nnz} vs {count_nonzero(ad.raw.X)}"
-
+                ad_x = ad.raw.X
+            h5ad_sum = sum(ad_x.data)  # calculate the actual sum of the X matrix values
+            assert raw_sum == h5ad_sum, error.format(eb.name, dataset.dataset_id, "sum", raw_sum, h5ad_sum)
+            h5ad_nnz = _count_nonzero(ad_x.data)
+            assert raw_nnz == h5ad_nnz, error.format(eb.name, dataset.dataset_id, "nnz", raw_nnz, h5ad_nnz)
+            h5ad_gene_count = np.count_nonzero(ad_x.sum(axis=0))
+            assert h5ad_gene_count == fdpm_gene_count, error.format(
+                eb.name, dataset.dataset_id, "gene_count", fdpm_gene_count, h5ad_gene_count
```

Comment:
nit: prefix or suffix test value variables with `expected`  and `actual`, for readability